main.go

```diff
@@ -21,54 +21,209 @@
+			log.Fatal(err)
+		}
 	}
 
 	opts := gen.Options{
-		PackageName: filepath.Base(outDir), // TODO: customize using flags
-		Output:      outFile,
+		OutputDir:     outputDir,
+		PackagePrefix: packagePrefix,
+		ThriftRoot:    thriftRoot,
+		NoRecurse:     noRecurse,
 	}
+
 	if err := gen.Generate(module, &opts); err != nil {
 		log.Fatal(err)
 	}
 }
+
+// verifyAncestry verifies that the Thrift file for the given module and the
+// Thrift files for all imported modules are contained within the directory
+// tree rooted at the given path.
+func verifyAncestry(m *compile.Module, root string) error {
+	return m.Walk(func(m *compile.Module) error {
+		path, err := filepath.Rel(root, m.ThriftPath)
+		if err != nil {
+			return fmt.Errorf(
+				"could not resolve path for %q: %v", m.ThriftPath, err)
+		}
+
+		if strings.HasPrefix(path, "..") {
+			return fmt.Errorf(
+				"%q is not contained in the %q directory tree",
+				m.ThriftPath, root)
+		}
+
+		return nil
+	})
+}
+
+// findCommonAncestor finds the deepest common ancestor for the given module
+// and all modules imported by it.
+func findCommonAncestor(m *compile.Module) (string, error) {
+	var result []string
+	var lastString string
+
+	err := m.Walk(func(m *compile.Module) error {
+		thriftPath := m.ThriftPath
+		if !filepath.IsAbs(thriftPath) {
+			return fmt.Errorf(
+				"ThriftPath must be absolute: %q is not absolute", thriftPath)
+		}
+
+		thriftDir := filepath.Dir(thriftPath)
+
+		// Split("/foo/bar", "/") = ["", "foo", "bar"]
+		parts := strings.Split(thriftDir, string(filepath.Separator))
+		if result == nil {
+			result = parts
+			lastString = thriftPath
+			return nil
+		}
+
+		result = commonPrefix(result, parts)
+		if len(result) == 1 && result[0] == "" {
+			return fmt.Errorf(
+				"%q does not share an ancestor with %q",
+				thriftPath, lastString)
+		}
+
+		lastString = thriftPath
+		return nil
+	})
+	if err != nil {
+		return "", err
+	}
+
+	if result[0] == "" {
+		// Join("", "foo") produces "foo", not "/foo"
+		result[0] = "/"
+	}
+
+	return filepath.Join(result...), nil
+}
+
+// commonPrefix finds the shortest common prefix for the two lists.
+//
+// An empty slice may be returned if the two lists don't have a common prefix.
+func commonPrefix(l, r []string) []string {
+	var i int
+	for i = 0; i < len(l) && i < len(r); i++ {
+		if l[i] != r[i] {
+			break
+		}
+	}
+	return l[:i]
+}
+
+// determinePackagePrefix determines the package prefix for Go packages
+// generated in this file.
+//
+// dir must be an absolute path.
+func determinePackagePrefix(dir string) (string, error) {
+	gopathList := os.Getenv("GOPATH")
+	if gopathList == "" {
+		return "", errors.New("$GOPATH is not set")
+	}
+
+	for _, gopath := range filepath.SplitList(gopathList) {
+		// entry in GOPATH might be a relative path
+		gopath, err := filepath.Abs(gopath)
+		if err != nil {
+			return "", err
+		}
+
+		packagePath, err := filepath.Rel(filepath.Join(gopath, "src"), dir)
+		if err != nil {
```

Comment:
don't think this is right, if `Rel` fails for one GOPATH, it may still succeed on another. you shouldn't be returning at that point. 

also update tests so that it only succeeds for the second path in `GOPATH` rather than the first every time.